ru/lessons/basics/functions.md

```diff
@@ -0,0 +1,216 @@
+
+Давайте рассмотрим простой пример:
+
+```elixir
+iex> sum = fn (a, b) -> a + b end
+iex> sum.(2, 3)
+5
+```
+
+### Краткий синтаксис
+
+Использование анонимных функций - настолько часто используемая возможность языка, что для нее есть специальное сокращение:
+
+```elixir
+iex> sum = &(&1 + &2)
+iex> sum.(2, 3)
+5
+```
+
+Как вы уже могли догадаться, в сокращенной версии параметры доступны как  `&1`, `&2`, `&3` и так далее.
+
+## Сопоставление с образцом
+
+Сопоставление с образцом в Elixir применяется не только для сопоставления переменных. Этот же инструмент используется в объявлении функций.
+
+Elixir использует сопоставление с образцом для определения первого подходящего набора параметров и вызывает соответствующую имплементацию:
+
+```elixir
+iex> handle_result = fn
+...>   {:ok, result} -> IO.puts "Handling result..."
+...>   {:error} -> IO.puts "An error has occurred!"
+...> end
+
+iex> some_result = 1
+iex> handle_result.({:ok, some_result})
+Handling result...
+
+iex> handle_result.({:error})
+An error has occurred!
+```
+
+## Именованные функции
+
+Можно определять именованные функции для дальнейшего их вызова по этим именам. Эти функции объявляются с помощью ключевого слова `def` в контексте модуля. Про модули будет подробнее рассказано в следующих уроках, в этом мы сосредоточимся только на именованных функциях.
+
+Функции, определенные в модуле, доступны из других модулей:
+
+```elixir
+defmodule Greeter do
+  def hello(name) do
+    "Hello, " <> name
+  end
+end
+
+iex> Greeter.hello("Sean")
+"Hello, Sean"
+```
+
+Если функция однострочная, то ее описание можно сократить с использованием `do:`:
+
+```elixir
+defmodule Greeter do
+  def hello(name), do: "Hello, " <> name
+end
+```
+
+Уже разобравшись в сопоставлении с образцом, давайте рассмотрим пример рекурсии с использованием именованных функций:
+
+```elixir
+defmodule Length do
+  def of([]), do: 0
+  def of([_|t]), do: 1 + of(t)
+end
+
+iex> Length.of []
+0
+iex> Length.of [1, 2, 3]
+3
+```
+
+### Закрытые функции
+
+Когда мы не хотим давать доступ к функции из других модулей, мы определяем закрытые (private) функции. Они могут быть вызваны только из этого же модуля. Такие функции определяются с помощью `defp`:
+
+```elixir
+defmodule Greeter do
+  def hello(name), do: phrase <> name
+  defp phrase, do: "Hello, "
+end
+
+iex> Greeter.hello("Sean")
+"Hello, Sean"
+
+iex> Greeter.phrase
+** (UndefinedFunctionError) undefined function: Greeter.phrase/0
+    Greeter.phrase()
+```
+
+### Ограничители
+
+Мы уже затрагивали ограничители в главе [Управляющие конструкции](../control-structures.md), теперь же рассмотрим их применение в именованных функциях. Ограничители проверяются только после того как только Elixir сопоставил функцию.
+
+В следующем примере у нас есть две функции с одинаковыми сигнатурами. Мы используем ограничители для определения какую из них использовать на основе типа аргумента:
```

Comment:
In this context (function definitions) - yes. https://ru.wikipedia.org/wiki/API#.D0.A1.D0.B8.D0.B3.D0.BD.D0.B0.D1.82.D1.83.D1.80.D0.B0_.D1.84.D1.83.D0.BD.D0.BA.D1.86.D0.B8.D0.B8

> Сигнатура функции — часть общего объявления функции, позволяющая средствам трансляции идентифицировать функцию среди других.